zcash_client_memory/src/types/account.rs

```diff
@@ -0,0 +1,780 @@
+        &mut self,
+        next_to_reserve: u32,
+    ) -> Result<Vec<(TransparentAddress, TransparentAddressMetadata)>, Error> {
+        if let Some(ephemeral_ivk) = self.ephemeral_ivk()? {
+            let first_unstored = self.first_unstored_index()?;
+            let range_to_store = first_unstored..(next_to_reserve.checked_add(GAP_LIMIT).unwrap());
+            if range_to_store.is_empty() {
+                return Ok(Vec::new());
+            }
+            return range_to_store
+                .map(|raw_index| {
+                    NonHardenedChildIndex::from_index(raw_index)
+                        .map(|address_index| {
+                            ephemeral_ivk
+                                .derive_ephemeral_address(address_index)
+                                .map(|addr| {
+                                    self.ephemeral_addresses.insert(
+                                        raw_index,
+                                        EphemeralAddress {
+                                            address: addr,
+                                            seen: None,
+                                            used: None,
+                                        },
+                                    );
+                                    (
+                                        addr,
+                                        TransparentAddressMetadata::new(
+                                            TransparentKeyScope::EPHEMERAL,
+                                            address_index,
+                                        ),
+                                    )
+                                })
+                        })
+                        .unwrap()
+                        .map_err(Into::into)
+                })
+                .collect::<Result<Vec<_>, _>>();
+        }
+        Ok(Vec::new())
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    pub fn mark_ephemeral_address_as_used(
+        &mut self,
+        address: &TransparentAddress,
+        tx_id: TxId,
+    ) -> Result<(), Error> {
+        // TODO: ephemeral_address_reuse_check
+        for (idx, addr) in self.ephemeral_addresses.iter_mut() {
+            if addr.address == *address {
+                addr.mark_used(tx_id);
+
+                // Maintain the invariant that the last `GAP_LIMIT` addresses are used and unseen.
+                let next_to_reserve = idx.checked_add(1).expect("ensured by constraint");
+                self.reserve_until(next_to_reserve)?;
+                return Ok(());
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "transparent-inputs")]
+    pub fn mark_ephemeral_address_as_seen(
+        &mut self,
+        // txns: &TransactionTable,
+        address: &TransparentAddress,
+        tx_id: TxId,
+    ) -> Result<(), Error> {
+        for (idx, addr) in self.ephemeral_addresses.iter_mut() {
+            if addr.address == *address {
+                // TODO: this
```

Comment:
Please create an issue for this TODO item if this is something out of the scope of this PR